### PR TITLE
fix: test race condition

### PR DESCRIPTION
**Pull request by Integralist**
We're seeing data race issues in CI:
https://github.com/domainr/epp2/actions/runs/8330541688/job/22795438262?pr=22 

There are a few different asynchronous layers within these epp2 tests...

- `echoServer` is run in a goroutine and reads from `t`.
- `echoServer` also runs a separate goroutine that writes to `t`.
- The _caller_ of `echoServer` also runs multiple goroutines to write to `t`.

The changes in this PR should resolve the data race issue. 

Instead of passing around the `*testing.T` I'm using channels to handle the coordination/synchronization of error messages.

### CODE REVIEW

File: protocol/dataunit/client_server_test.go [hunk 1]
```diff
@@ -56,33 +81,43 @@ func TestClientContextDeadline(t *testing.T) {
...
 
 	wantErr := errors.New("test deadline exceeded")
 	ctx, cancel2 := context.WithDeadlineCause(ctx, time.Now(), wantErr)
 	defer cancel2()
 
 	_, err := c.ExchangeDataUnit(ctx, []byte("hello"))
-	if err != wantErr {
-		t.Errorf("ExchangeDataUnit(): err == %v, expected %v", err, wantErr)
-		t.Fail()
+	if !errors.Is(err, wantErr) {
```
Comment (cee-dub) [inline protocol/dataunit/client_server_test.go:93]: I have a feeling @ydnar  was writing this with TinyGo in mind and `errors.Is` is one of its main missing features these days.

File: protocol/dataunit/client_server_test.go [hunk 2]
```diff
@@ -94,15 +129,22 @@ func TestServerContextCancelled(t *testing.T) {
 	c := &Client{Conn: clientConn}
 	s := &Server{Conn: serverConn}
 
-	go c.ExchangeDataUnit(ctx, []byte("hello"))
+	go func() {
+		_, _ = c.ExchangeDataUnit(ctx, []byte("hello"))
+	}()
 
-	wantErr := errors.New("server context canceled")
+	wantErr := errors.New("server context cancelled")
```
Comment (cee-dub) [inline protocol/dataunit/client_server_test.go:136]: 🚲🏢 (huge bike shed warning) ⚠️ ⛔️:

Turns out both spellings are correct and we have chosen the shorter one in Domainr's codebase.

File: protocol/dataunit/client_server_test.go [hunk 3]
```diff
@@ -153,28 +199,31 @@ func echoServer(t *testing.T, ctx context.Context, s *Server) error {
...
-				t.Errorf("echoServer: WriteDataUnit(): err == %v", err)
+				echoServerErr <- fmt.Errorf("echoServer: RespondDataUnit(): err == %w", err)
 			}
 		}()
 	}
 }
 
 func randDuration(max time.Duration) time.Duration {
-	return time.Duration(rand.Int63n(int64(max)))
+	return time.Duration(rand.Int64N(int64(max)))
```
Comment (cee-dub) [inline protocol/dataunit/client_server_test.go:223]: Use `rand.N` and avoid all the type conversions. https://pkg.go.dev/math/rand/v2#example-N

**Review by cee-dub — APPROVED** [review]
The only possibly-contentious item is the error comparison that is known not to be compatible with TinyGo.

IMO, we should develop this as if TinyGo will gain that feature by the time Domainr and TinyGo need to ship these things together.

File: protocol/dataunit/client_server_test.go (hunk 1 above)
Comment (Integralist) [reply protocol/dataunit/client_server_test.go:93]: Ah! OK fair enough I'll undo those changes 👍🏻

File: protocol/dataunit/client_server_test.go (hunk 2 above)
Comment (Integralist) [reply protocol/dataunit/client_server_test.go:136]: lol, I clearly don't think any other country exists outside of the UK 😄 🤦🏻 

I'll revert that.

File: protocol/dataunit/client_server_test.go (hunk 3 above)
Comment (Integralist) [reply protocol/dataunit/client_server_test.go:223]: Cool, will do that 👍🏻

**Comment by Integralist** [comment]
All changes have been made. I'm stepping away from the keyboard for 30mins and I'll merge when I'm back.

**Pull request merged**